File: pkg/ir/feature_column.go
// Copyright 2020 The SQLFlow Authors. All rights reserved.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package ir

import "fmt"

// FeatureColumn corresponds to the COLUMN clause in TO TRAIN.
type FeatureColumn interface {
	GetFieldDesc() []*FieldDesc
	ApplyTo(*FieldDesc) (FeatureColumn, error)
}

// CategoryColumn corresponds to categorical column
type CategoryColumn interface {
	FeatureColumn
	NumClass() int64
}

// FieldDesc describes a field used as the input to a feature column.
type FieldDesc struct {
	Name      string `json:"name"`      // the name for a field, e.g. "petal_length"
	DType     int    `json:"dtype"`     // e.g. "float", "int32"
	Delimiter string `json:"delimiter"` // Needs to be "," if the field saves strings like "1,23,42".
	Format    string `json:format`      // The data format, "csv" or "libsvm"
	Shape     []int  `json:"shape"`     // [3] if the field saves strings of three numbers like "1,23,42".
	IsSparse  bool   `json:"is_sparse"` // If the field saves a sparse tensor.
	// Vocabulary stores all possible enumerate values if the column type is string,
	// e.g. the column values are: "MALE", "FEMALE", "NULL"
	Vocabulary map[string]string `json:"vocabulary"` // use a map to generate a list without duplication
	// if the column data is used as embedding(category_column()), the `num_buckets` should use the maxID
	// appeared in the sample data. if error still occurs, users should set `num_buckets` manually.
	MaxID int64
}

// Possible DType values in FieldDesc
const (
	Int int = iota
	Float
	String
)

// NumericColumn represents a dense tensor for the model input
//
// FieldDesc indicates the meta information for decoding the field. Please be aware
// that FieldDesc also contains information for dimension and data type
type NumericColumn struct {
	FieldDesc *FieldDesc
}

// GetFieldDesc returns FieldDesc member
func (c *NumericColumn) GetFieldDesc() []*FieldDesc {
	return []*FieldDesc{c.FieldDesc}
}

// ApplyTo applies the FeatureColumn to a new field
func (c *NumericColumn) ApplyTo(other *FieldDesc) (FeatureColumn, error) {
	return &NumericColumn{other}, nil
}

// BucketColumn represents `tf.feature_column.bucketized_column`
// ref: https://www.tensorflow.org/api_docs/python/tf/feature_column/bucketized_column
type BucketColumn struct {
	SourceColumn *NumericColumn
	Boundaries   []int
}

// GetFieldDesc returns FieldDesc member
func (c *BucketColumn) GetFieldDesc() []*FieldDesc {
	return c.SourceColumn.GetFieldDesc()
}

// ApplyTo applies the FeatureColumn to a new field
func (c *BucketColumn) ApplyTo(other *FieldDesc) (FeatureColumn, error) {
	sourceColumn, err := c.SourceColumn.ApplyTo(other)
	if err != nil {
		return nil, err
	}
	return &BucketColumn{
		SourceColumn: sourceColumn.(*NumericColumn),
		Boundaries:   c.Boundaries,
	}, nil
}

// NumClass returns class number of BucketColumn
func (c *BucketColumn) NumClass() int64 {
	return int64(len(c.Boundaries)) + 1
}

// CrossColumn represents `tf.feature_column.crossed_column`
// ref: https://www.tensorflow.org/api_docs/python/tf/feature_column/crossed_column
type CrossColumn struct {
	Keys           []interface{}
	HashBucketSize int64
}

// GetFieldDesc returns FieldDesc member
func (c *CrossColumn) GetFieldDesc() []*FieldDesc {
	var retKeys []*FieldDesc
	for idx, k := range c.Keys {
		if strKey, ok := k.(string); ok {
			retKeys = append(retKeys, &FieldDesc{
				Name:  strKey,
				DType: String,
				Shape: []int{1},
			})
		} else if _, ok := k.(FeatureColumn); ok {
			retKeys = append(retKeys, c.Keys[idx].(*NumericColumn).GetFieldDesc()[0])
		}
		// k is not possible to be neither string and FeatureColumn, the ir_generator should
		// catch the syntax error.
	}
	return retKeys
}

// ApplyTo applies the FeatureColumn to a new field
func (c *CrossColumn) ApplyTo(other *FieldDesc) (FeatureColumn, error) {
	return nil, fmt.Errorf("CrossColumn doesn't support the method ApplyTo")
}

// NumClass returns class number of CrossColumn
func (c *CrossColumn) NumClass() int64 {
	return c.HashBucketSize
}

// CategoryIDColumn represents `tf.feature_column.categorical_column_with_identity`
// ref: https://www.tensorflow.org/api_docs/python/tf/feature_column/categorical_column_with_identity
type CategoryIDColumn struct {
	FieldDesc  *FieldDesc
	BucketSize int64
}

// GetFieldDesc returns FieldDesc member
func (c *CategoryIDColumn) GetFieldDesc() []*FieldDesc {
	return []*FieldDesc{c.FieldDesc}
}

// ApplyTo applies the FeatureColumn to a new field
func (c *CategoryIDColumn) ApplyTo(other *FieldDesc) (FeatureColumn, error) {
	return &CategoryIDColumn{other, c.BucketSize}, nil
}

// NumClass returns class number of CategoryIDColumn
func (c *CategoryIDColumn) NumClass() int64 {
	return c.BucketSize
}

// CategoryHashColumn represents `tf.feature_column.categorical_column_with_hash_bucket`
// ref: https://www.tensorflow.org/api_docs/python/tf/feature_column/categorical_column_with_hash_bucket
type CategoryHashColumn struct {
	FieldDesc  *FieldDesc
	BucketSize int64
}

// GetFieldDesc returns FieldDesc member
func (c *CategoryHashColumn) GetFieldDesc() []*FieldDesc {
	return []*FieldDesc{c.FieldDesc}
}

// ApplyTo applies the FeatureColumn to a new field
func (c *CategoryHashColumn) ApplyTo(other *FieldDesc) (FeatureColumn, error) {
	return &CategoryHashColumn{other, c.BucketSize}, nil
}

// NumClass returns class number of CategoryHashColumn
func (c *CategoryHashColumn) NumClass() int64 {
	return c.BucketSize
}

// SeqCategoryIDColumn represents `tf.feature_column.sequence_categorical_column_with_identity`
// ref: https://www.tensorflow.org/api_docs/python/tf/feature_column/sequence_categorical_column_with_identity
type SeqCategoryIDColumn struct {
	FieldDesc  *FieldDesc
	BucketSize int64
}

// GetFieldDesc returns FieldDesc member
func (c *SeqCategoryIDColumn) GetFieldDesc() []*FieldDesc {
	return []*FieldDesc{c.FieldDesc}
}

// ApplyTo applies the FeatureColumn to a new field
func (c *SeqCategoryIDColumn) ApplyTo(other *FieldDesc) (FeatureColumn, error) {
	return &SeqCategoryIDColumn{other, c.BucketSize}, nil
}

// NumClass returns class number of SeqCategoryIDColumn
func (c *SeqCategoryIDColumn) NumClass() int64 {
	return c.BucketSize
}

// EmbeddingColumn represents `tf.feature_column.embedding_column`
// ref: https://www.tensorflow.org/api_docs/python/tf/feature_column/embedding_column
type EmbeddingColumn struct {
	CategoryColumn
	Dimension   int
	Combiner    string
	Initializer string
	// only used when EMBEDDING(col_name, ...) this will set CategoryColumn = nil
	// will fill the feature column details using feature_derivation
	Name string
}

// GetFieldDesc returns FieldDesc member
func (c *EmbeddingColumn) GetFieldDesc() []*FieldDesc {
	if c.CategoryColumn == nil {
		return []*FieldDesc{}
	}
	return c.CategoryColumn.(FeatureColumn).GetFieldDesc()
}

// ApplyTo applies the FeatureColumn to a new field
func (c *EmbeddingColumn) ApplyTo(other *FieldDesc) (FeatureColumn, error) {
	ret := &EmbeddingColumn{
		Dimension:   c.Dimension,
		Combiner:    c.Combiner,
		Initializer: c.Initializer,
		Name:        other.Name,
	}
	if c.CategoryColumn != nil {
		var err error
		fc, err := c.CategoryColumn.ApplyTo(other)
		if err != nil {
			return nil, err
		}

		if categoryFc, ok := fc.(CategoryColumn); !ok {
			ret.CategoryColumn = categoryFc
		} else {
			return nil, fmt.Errorf("invalid EmbeddingColumn.ApplyTo return value")
		}
	}
	return ret, nil
}

// IndicatorColumn represents `tf.feature_column.indicator_column`
// ref: https://www.tensorflow.org/api_docs/python/tf/feature_column/indicator_column
type IndicatorColumn struct {
	CategoryColumn FeatureColumn
	// only used when INDICATOR(col_name, ...) this will set CategoryColumn = nil
	// will fill the feature column details using feature_derivation
	Name string
}

// GetFieldDesc returns FieldDesc member
func (c *IndicatorColumn) GetFieldDesc() []*FieldDesc {
	if c.CategoryColumn == nil {
		return []*FieldDesc{}
	}
	return c.CategoryColumn.(FeatureColumn).GetFieldDesc()
}

// ApplyTo applies the FeatureColumn to a new field
func (c *IndicatorColumn) ApplyTo(other *FieldDesc) (FeatureColumn, error) {
	ret := &IndicatorColumn{Name: other.Name}
	if c.CategoryColumn != nil {
		var err error
		ret.CategoryColumn, err = c.CategoryColumn.ApplyTo(other)
		if err != nil {
			return nil, err
		}
	}
	return ret, nil
}

File: python/runtime/feature_derivation/feature_column.py
# Copyright 2020 The SQLFlow Authors. All rights reserved.
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json

import six

INT = 0
FLOAT = 1
STRING = 2


class FieldDesc(object):
    """
    FieldDesc describes a field used as the input to a feature column.

    Args:
        name (str): the field name. Default "".
        dtype (int): the data type of the field. It must be one of INT,
            FLOAT and STRING. Default INT.
        delimiter (str): the delimiter of the field data. Default "".
        format (str): the format of the field data. Default "".
        shape (list[int]): the shape of the field data. Default None.
        is_sparse (bool): whether the field data is sparse. Default False.
        vocabulary (list[str]): the vocabulary used for categorical feature column. Default None.
        max_id (int): the maximum id number of the field data. Used in CategoryIDColumn. Default 0.
    """
    def __init__(self,
                 name="",
                 dtype=INT,
                 delimiter="",
                 format="",
                 shape=None,
                 is_sparse=False,
                 vocabulary=None,
                 max_id=0):
        self.name = name
        self.dtype = dtype
        self.delimiter = delimiter
        self.format = format
        self.shape = shape
        self.is_sparse = is_sparse
        self.vocabulary = vocabulary
        self.max_id = max_id

    def to_json(self):
        """
        Convert the FieldDesc object to a json string.

        Returns:
            A string which represents the json value of the FieldDesc object.
        """
        return json.dumps({
            "name": self.name,
            "dtype": self.dtype,
            "delimiter": self.delimiter,
            "format": self.format,
            "shape": self.shape,
            "is_sparse": self.is_sparse,
            "vocabulary": self.vocabulary,
            "max_id": self.max_id,
        })


class FeatureColumn(object):
    """
    FeatureColumn corresponds to the COLUMN clause in the TO TRAIN statement.
    It is the base class of all feature column classes.
    """
    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the feature
        column object holds.

        Returns:
            A list of the FieldDesc objects.
        """
        raise NotImplementedError()

    def apply_to(self, field_desc):
        """
        Create a new feature column object of the same type
        that holds the given FieldDesc object.

        Args:
            field_desc (FieldDesc): the given FieldDesc object.

        Returns:
            A new feature column object which is of the same type,
            and holds the given FieldDesc object.
        """
        raise NotImplementedError()


class CategoryColumn(FeatureColumn):
    """
    CategoryColumn corresponds to the categorical feature column.
    It is the base class of all categorical feature column classes.
    """
    def num_class(self):
        """
        Get the class number of the categorical feature column.

        Returns:
            An integer which represents the class number.
        """
        raise NotImplementedError()


class NumericColumn(FeatureColumn):
    """
    NumericColumn represents a dense or sparse numeric feature.

    Args:
        field_desc (FieldDesc): the underlying FieldDesc object that the
            NumericColumn object holds.
    """
    def __init__(self, field_desc):
        assert isinstance(field_desc, FieldDesc)
        self.field_desc = field_desc

    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the NumericColumn
        object holds.

        Returns:
            A list of the FieldDesc objects, whose length is 1.
        """
        return [self.field_desc]

    def apply_to(self, field_desc):
        """
        Create a new NumericColumn object that holds the given
        FieldDesc object.

        Args:
            field_desc (FieldDesc): the given FieldDesc object.

        Returns:
            A new NumericColumn object which holds the given
            FieldDesc object.
        """
        return NumericColumn(field_desc)


class BucketColumn(CategoryColumn):
    """
    BucketColumn represents a bucketized feature column.

    Args:
        source_column (NumericColumn): the underlying NumericColumn object.
        boundaries (list[int|float]): the boundaries of the buckets.
    """
    def __init__(self, source_column, boundaries):
        assert isinstance(
            source_column,
            NumericColumn), "source_column of BUCKET must be of numeric type"
        self.source_column = source_column
        self.boundaries = boundaries

    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the BucketColumn
        object holds.

        Returns:
            A list of the FieldDesc objects, whose length is 1.
        """
        return self.source_column.get_field_desc()

    def apply_to(self, field_desc):
        """
        Create a new BucketColumn object that holds the given
        FieldDesc object.

        Args:
            field_desc (FieldDesc): the given FieldDesc object.

        Returns:
            A new BucketColumn object which holds the given
            FieldDesc object.
        """
        source_column = self.source_column.apply_to(field_desc)
        return BucketColumn(source_column, self.boundaries)

    def num_class(self):
        """
        Get the class number of the BucketColumn object.

        Returns:
            An integer which is equal to len(boundaries) + 1.
        """
        return len(self.boundaries) + 1


class CategoryIDColumn(CategoryColumn):
    """
    CategoryIDColumn represents a categorical id feature column.

    Args:
        field_desc (FieldDesc): the underlying FieldDesc object.
        bucket_size (int): the bucket size.
    """
    def __init__(self, field_desc, bucket_size):
        assert isinstance(field_desc, FieldDesc)
        self.field_desc = field_desc
        self.bucket_size = bucket_size

    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the CategoryIDColumn
        object holds.

        Returns:
            A list of the FieldDesc objects, whose length is 1.
        """
        return [self.field_desc]

    def apply_to(self, field_desc):
        """
        Create a new CategoryIDColumn object that holds the given
        FieldDesc object.

        Args:
            field_desc (FieldDesc): the given FieldDesc object.

        Returns:
            A new CategoryIDColumn object which holds the given
            FieldDesc object.
        """
        return CategoryIDColumn(field_desc, self.bucket_size)

    def num_class(self):
        """
        Get the class number of the CategoryIDColumn object.

        Returns:
            An integer which is equal to bucket_size.
        """
        return self.bucket_size


class CategoryHashColumn(CategoryIDColumn):
    """
    CategoryHashColumn represents a categorical hash feature column.

    Args:
        field_desc (FieldDesc): the underlying FieldDesc object.
        bucket_size (int): the bucket size for hashing.
    """
    def __init__(self, field_desc, bucket_size):
        assert isinstance(field_desc, FieldDesc)
        self.field_desc = field_desc
        self.bucket_size = bucket_size

    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the CategoryHashColumn
        object holds.

        Returns:
            A list of the FieldDesc objects, whose length is 1.
        """
        return [self.field_desc]

    def apply_to(self, field_desc):
        """
        Create a new CategoryHashColumn object that holds the given
        FieldDesc object.

        Args:
            field_desc (FieldDesc): the given FieldDesc object.

        Returns:
            A new CategoryHashColumn object which holds the given
            FieldDesc object.
        """
        return CategoryHashColumn(field_desc, self.bucket_size)

    def num_class(self):
        """
        Get the class number of the CategoryHashColumn object.

        Returns:
            An integer which is equal to bucket_size.
        """
        return self.bucket_size


class SeqCategoryIDColumn(CategoryIDColumn):
    """
    SeqCategoryIDColumn represents a sequential categorical id feature column.

    Args:
        field_desc (FieldDesc): the underlying FieldDesc object.
        bucket_size (int): the bucket size.
    """
    def __init__(self, field_desc, bucket_size):
        assert isinstance(field_desc, FieldDesc)
        self.field_desc = field_desc
        self.bucket_size = bucket_size

    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the SeqCategoryIDColumn
        object holds.

        Returns:
            A list of the FieldDesc objects, whose length is 1.
        """
        return [self.field_desc]

    def apply_to(self, field_desc):
        """
        Create a new SeqCategoryIDColumn object that holds the given
        FieldDesc object.

        Args:
            field_desc (FieldDesc): the given FieldDesc object.

        Returns:
            A new SeqCategoryIDColumn object which holds the given
            FieldDesc object.
        """
        return SeqCategoryIDColumn(field_desc, self.bucket_size)

    def num_class(self):
        """
        Get the class number of the SeqCategoryIDColumn object.

        Returns:
            An integer which is equal to bucket_size.
        """
        return self.bucket_size


class CrossColumn(CategoryColumn):
    """
    CrossColumn represents a crossed feature column.

    Args:
        keys (str|NumericColumn): the underlying feature column name or NumericColumn object.
        hash_bucket_size (int): the bucket size for hashing.
    """
    def __init__(self, keys, hash_bucket_size):
        for k in keys:
            assert isinstance(k, six.string_types) or isinstance(k, NumericColumn), \
                "keys of CROSS must be of either string or numeric type"

        self.keys = keys
        self.hash_bucket_size = hash_bucket_size

    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the CrossColumn
        object holds.

        Returns:
            A list of the FieldDesc objects.
        """
        descs = []
        for k in self.keys:
            if isinstance(k, six.string_types):
                descs.append(FieldDesc(name=k, dtype=STRING, shape=[1]))
            elif isinstance(k, NumericColumn):
                descs.extend(k.get_field_desc())
            else:
                raise ValueError("unsupported type %s" % type(k))

        return descs

    def apply_to(self, field_desc):
        """
        This method is not supported for the CrossColumn object.
        """
        raise NotImplementedError("CROSS does not support apply_to method")

    def num_class(self):
        """
        Get the class number of the CrossColumn object.

        Returns:
            An integer which is equal to hash_bucket_size.
        """
        return self.hash_bucket_size


class EmbeddingColumn(FeatureColumn):
    """
    EmbeddingColumn represents an embedding feature column.

    Args:
        category_column (CategoryColumn): the underlying CategoryColumn object.
        dimension (int): the dimension of the embedding.
        combiner (str): how to reduce if there are multiple entries in a single row.
            Currently 'mean', 'sqrtn' and 'sum' are supported.
        initializer (str): the initializer of the embedding table.
        name (str): only used when category_column=None. In this case, the
            category_column would be filled automaticaly in the feature derivation
            stage.
    """
    def __init__(self,
                 category_column=None,
                 dimension=0,
                 combiner="",
                 initializer="",
                 name=""):
        if category_column is not None:
            assert isinstance(category_column, CategoryColumn)

        self.category_column = category_column
        self.dimension = dimension
        self.combiner = combiner
        self.initializer = initializer
        self.name = name

    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the EmbeddingColumn
        object holds.

        Returns:
            A list of the FieldDesc objects.
        """
        if self.category_column is None:
            return [FieldDesc()]

        return self.category_column.get_field_desc()

    def apply_to(self, field_desc):
        """
        Create a new EmbeddingColumn object that holds the given
        FieldDesc object.

        Args:
            field_desc (FieldDesc): the given FieldDesc object.

        Returns:
            A new EmbeddingColumn object which holds the given
            FieldDesc object.
        """
        if self.category_column is not None:
            category_column = self.category_column.apply_to(field_desc)
        else:
            category_column = None

        return EmbeddingColumn(category_column=category_column,
                               dimension=self.dimension,
                               combiner=self.combiner,
                               initializer=self.initializer,
                               name=self.name)


class IndicatorColumn(FeatureColumn):
    """
    IndicatorColumn represents the one-hot feature column.

    Args:
        category_column (CategoryColumn): the underlying CategoryColumn object.
        name (str): only used when category_column=None. In this case, the
            category_column would be filled automaticaly in the feature derivation
            stage.
    """
    def __init__(self, category_column=None, name=""):
        if category_column is not None:
            assert isinstance(category_column, CategoryColumn)

        self.category_column = category_column
        self.name = name

    def get_field_desc(self):
        """
        Get the underlying FieldDesc object list that the IndicatorColumn
        object holds.

        Returns:
            A list of the FieldDesc objects.
        """
        if self.category_column is None:
            return [FieldDesc()]

        return self.category_column.get_field_desc()

    def apply_to(self, field_desc):
        """
        Create a new IndicatorColumn object that holds the given
        FieldDesc object.

        Args:
            field_desc (FieldDesc): the given FieldDesc object.

        Returns:
            A new IndicatorColumn object which holds the given
            FieldDesc object.
        """
        if self.category_column is not None:
            category_column = self.category_column.apply_to(field_desc)
        else:
            category_column = None

        return IndicatorColumn(category_column, self.name)
